aws/security_hub.go

```diff
@@ -0,0 +1,182 @@
+		telemetry.TrackEvent(commonTelemetry.EventContext{
+			EventName: "Error disassociating members from security hub",
+		}, map[string]interface{}{
+			"region": *svc.Config.Region,
+			"reason": "Unable to disassociate",
+		})
+		return
+	}
+	logging.Logger.Debugf("%d member accounts disassociated", len(accountIds))
+
+	// Once disassociated, member accounts can be deleted
+	_, err = svc.DeleteMembers(&securityhub.DeleteMembersInput{AccountIds: accountIds})
+	if err != nil {
+		logging.Logger.Errorf("[Failed] Failed to delete members from security hub")
+		telemetry.TrackEvent(commonTelemetry.EventContext{
+			EventName: "Error deleting members from security hub",
+		}, map[string]interface{}{
+			"region": *svc.Config.Region,
+			"reason": "Unable to delete",
+		})
+		return
+	}
+	logging.Logger.Debugf("%d member accounts deleted", len(accountIds))
+}
+
+func disassociateAdministratorAccount(svc *securityhub.SecurityHub) {
+
+	// Check for an associated admin account as it must be disassociated before disabling security hub
+	adminAccount, err := svc.GetAdministratorAccount(&securityhub.GetAdministratorAccountInput{})
+	if err != nil {
+		logging.Logger.Errorf("[Failed] Failed to check for administrator account")
+		telemetry.TrackEvent(commonTelemetry.EventContext{
+			EventName: "Error checking for administrator account in security hub",
+		}, map[string]interface{}{
+			"region": *svc.Config.Region,
+			"reason": "Unable to find admin account",
+		})
+	}
+	// Disassociate if one exists
+	if adminAccount.Administrator != nil {
+		_, err := svc.DisassociateFromAdministratorAccount(&securityhub.DisassociateFromAdministratorAccountInput{})
+		if err != nil {
+			logging.Logger.Errorf("[Failed] Failed to disassociate from administrator account")
+			telemetry.TrackEvent(commonTelemetry.EventContext{
+				EventName: "Error disassociating administrator account in security hub",
+			}, map[string]interface{}{
+				"region": *svc.Config.Region,
+				"reason": "Unable to disassociate admin account",
+			})
+		} else {
+			logging.Logger.Debugf("Successfully disassociated from administrator account")
+		}
+	}
+}
+
+func nukeSecurityHub(session *session.Session, securityHubArns []string) error {
+	svc := securityhub.New(session)
+
+	// Check for and remove any member accounts in security hub
+	// Security Hub cannot be disabled with active member accounts
+	memberAccountIds, err := getAllSecurityHubMembers(svc)
+	if err != nil {
+		telemetry.TrackEvent(commonTelemetry.EventContext{
+			EventName: "Error finding security hub member accounts",
+		}, map[string]interface{}{
+			"region": *svc.Config.Region,
+			"reason": "Error finding security hub member accounts",
+		})
+	}
+	if err == nil && len(memberAccountIds) > 0 {
+		removeMembersFromHub(svc, memberAccountIds)
+	}
+
+	// Disassociate account from any administrator accounts
+	// Security hub cannot be disable with an active administrator account
+	disassociateAdministratorAccount(svc)
```

Comment:
What happens to an error that occurs in `disassociateAdministratorAccount`? Would it be propagated back up to stop execution from the main loop in `aws.go`?

Reply:
I've refactored some of these functions